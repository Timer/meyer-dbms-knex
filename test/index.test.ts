import Knex, { PgConnectionConfig } from 'knex';
import Meyer from 'meyer';
import path from 'path';
import KnexDbms from '../src';

const config: PgConnectionConfig = {
  host: 'localhost',
  port: 26043,
  user: 'sa',
  password: 'Meyer#123',
  database: 'testing',
};
const dbms = new KnexDbms('pg', config);

let testingKnex: Knex;
beforeAll(async () => {
  const boostrapConfig: PgConnectionConfig = {
    ...config,
    database: 'bootstrap',
  };

  const setupKnex = Knex({
    client: 'pg',
    connection: boostrapConfig,
    pool: { min: 0, max: 1, idleTimeoutMillis: 100, reapIntervalMillis: 100 },
  });

  await setupKnex.raw(`DROP DATABASE IF EXISTS "testing";`);
  await setupKnex.raw(`CREATE DATABASE "testing"`);

  await setupKnex.destroy();

  testingKnex = Knex({
    client: 'pg',
    connection: config,
    pool: { min: 0, max: 1, idleTimeoutMillis: 100, reapIntervalMillis: 100 },
  });
});

afterAll(async () => {
  if (testingKnex) await testingKnex.destroy();
  await dbms.close();
});

describe('Knex Provider', () => {
  test('it runs migrations v1', async () => {
    const m = new Meyer({
      dbms,
      migrationsPath: path.resolve(__dirname, 'migrations-v1'),
    });
    await m.execute();

    expect(await testingKnex.table(`abc`)).toMatchSnapshot();
  });
  test('it runs migrations v2', async () => {
    const m = new Meyer({
      dbms,
      migrationsPath: path.resolve(__dirname, 'migrations-v2'),
    });
    await m.execute();
    expect(await testingKnex.table(`abc`)).toMatchSnapshot();
    expect(await testingKnex.table(`def`)).toMatchSnapshot();
  });
  test('it fails migrations v3 in prod', async () => {
    const m = new Meyer({
      dbms,
      migrationsPath: path.resolve(__dirname, 'migrations-v3'),
    });
    let err;
    try {
      await m.execute();
    } catch (e) {
      err = e;
    }
    expect(err).toMatchSnapshot();
    expect(await testingKnex.table(`abc`)).toMatchSnapshot();
    expect(await testingKnex.table(`def`)).toMatchSnapshot();
  });
  test('it runs migrations v3 in dev', async () => {
    const m = new Meyer({
      dbms,
      migrationsPath: path.resolve(__dirname, 'migrations-v3'),
      development: true,
    });
    await m.execute();
    expect(await testingKnex.table(`abc`)).toMatchSnapshot();
    expect(await testingKnex.table(`jow`)).toMatchSnapshot();
    expect(testingKnex.table(`def`)).rejects.toThrowErrorMatchingInlineSnapshot(
      `"select * from \\"def\\" - relation \\"def\\" does not exist"`
    );
  });
});
