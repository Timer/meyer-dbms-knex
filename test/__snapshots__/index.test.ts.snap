// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Knex Provider it fails migrations v3 in prod 1`] = `[Error: ... migration 2 has a hash mismatch, aborting ...]`;

exports[`Knex Provider it fails migrations v3 in prod 2`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
  Object {
    "id": 3,
  },
]
`;

exports[`Knex Provider it fails migrations v3 in prod 3`] = `
Array [
  Object {
    "id": 4,
  },
  Object {
    "id": 5,
  },
  Object {
    "id": 6,
  },
]
`;

exports[`Knex Provider it runs migrations v1 1`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
  Object {
    "id": 3,
  },
]
`;

exports[`Knex Provider it runs migrations v2 1`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
  Object {
    "id": 3,
  },
]
`;

exports[`Knex Provider it runs migrations v2 2`] = `
Array [
  Object {
    "id": 4,
  },
  Object {
    "id": 5,
  },
  Object {
    "id": 6,
  },
]
`;

exports[`Knex Provider it runs migrations v3 in dev 1`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
  Object {
    "id": 3,
  },
]
`;

exports[`Knex Provider it runs migrations v3 in dev 2`] = `
Array [
  Object {
    "id": 7,
  },
  Object {
    "id": 8,
  },
  Object {
    "id": 9,
  },
  Object {
    "id": 10,
  },
]
`;
